smserver/smserver_test.go

```diff
@@ -515,3 +661,89 @@ func (client *testClient) close() {
 		client.connection.Close()
 	}
 }
+
+func checkLogs(notification pb.SMService_SubscribeSMNotificationsClient, testLogs []testLogData) (err error) {
+	notificationChan := make(chan *pb.SMNotifications, 1)
+
+	go func() {
+		for {
+			smData, _ := notification.Recv()
+			notificationChan <- smData
+		}
+	}()
+
+	var currentIndex int
+
+	for {
+		select {
+		case rawLog := <-notificationChan:
+			logData := rawLog.GetLog()
+			if logData == nil {
+				return errIncorrectType
+			}
+
+			if !proto.Equal(logData, &testLogs[currentIndex].expectedPbLog) {
+				return aoserrors.New("received log != send log")
+			}
+
+			currentIndex++
+
+			if currentIndex >= len(testLogs) {
+				return nil
+			}
+
+		case <-time.After(5 * time.Second):
+			return errTimeout
+		}
+	}
+}
+
+func compareServiceLogRequests(expected []pb.InstanceLogRequest, received []cloudprotocol.RequestServiceLog) error {
+	if len(expected) != len(received) {
+		return aoserrors.New("incorrect len of requests")
+	}
+
+	for i := range expected {
+		if expected[i].LogId != received[i].LogID {
+			return aoserrors.New("incorrect LogID")
+		}
+
+		if received[i].From == nil {
+			if expected[i].From != nil {
+				return aoserrors.New("incorrect from timestamp")
+			}
+		} else if *received[i].From != expected[i].From.AsTime() {
```

Comment:
It will crash if expected[i].From == nil.

Reply:
fixed